File: Packages/OpenVibbleKit/Sources/BuddyPersona/PersonaManifest.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

import Foundation

public struct PersonaManifest: Codable, Sendable, Equatable {
    public let name: String
    public let mode: Mode
    public let colors: Palette
    public let states: [String: StateFrames]

    public enum Mode: String, Codable, Sendable { case gif, text }

    public struct Palette: Codable, Sendable, Equatable {
        public let body: String?
        public let bg: String?
        public let text: String?
        public let textDim: String?
        public let ink: String?

        public init(body: String? = nil, bg: String? = nil, text: String? = nil, textDim: String? = nil, ink: String? = nil) {
            self.body = body
            self.bg = bg
            self.text = text
            self.textDim = textDim
            self.ink = ink
        }
    }

    public init(name: String, mode: Mode = .gif, colors: Palette = Palette(), states: [String: StateFrames]) {
        self.name = name
        self.mode = mode
        self.colors = colors
        self.states = states
    }

    enum CodingKeys: String, CodingKey { case name, mode, colors, states }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.name = try container.decode(String.self, forKey: .name)
        self.mode = (try? container.decode(Mode.self, forKey: .mode)) ?? .gif
        self.colors = (try? container.decode(Palette.self, forKey: .colors)) ?? Palette()
        self.states = (try? container.decode([String: StateFrames].self, forKey: .states)) ?? [:]
    }

    public func frames(for slug: String) -> StateFrames? {
        states[slug]
    }
}

public enum StateFrames: Codable, Sendable, Equatable {
    case single(String)
    case variants([String])
    case text(frames: [String], delayMs: Int)

    public var filenames: [String] {
        switch self {
        case .single(let name): return [name]
        case .variants(let names): return names
        case .text: return []
        }
    }

    public init(from decoder: Decoder) throws {
        // Try string
        if let single = try? decoder.singleValueContainer().decode(String.self) {
            self = .single(single); return
        }
        // Try array of strings
        if let array = try? decoder.singleValueContainer().decode([String].self) {
            self = .variants(array); return
        }
        // Try object { frames: [String], delay: Int }
        let obj = try decoder.container(keyedBy: ObjectKeys.self)
        let frames = try obj.decode([String].self, forKey: .frames)
        let delay = (try? obj.decode(Int.self, forKey: .delayMs))
            ?? (try? obj.decode(Int.self, forKey: .delay))
            ?? 200
        self = .text(frames: frames, delayMs: delay)
    }

    public func encode(to encoder: Encoder) throws {
        switch self {
        case .single(let s):
            var c = encoder.singleValueContainer()
            try c.encode(s)
        case .variants(let arr):
            var c = encoder.singleValueContainer()
            try c.encode(arr)
        case .text(let frames, let delay):
            var c = encoder.container(keyedBy: ObjectKeys.self)
            try c.encode(frames, forKey: .frames)
            try c.encode(delay, forKey: .delayMs)
        }
    }

    private enum ObjectKeys: String, CodingKey {
        case frames
        case delayMs = "delay_ms"
        case delay
    }
}

File: android/buddy-persona/src/main/kotlin/com/openvibble/persona/PersonaManifest.kt
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

package com.openvibble.persona

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive

@Serializable
data class PersonaPalette(
    val body: String? = null,
    val bg: String? = null,
    val text: String? = null,
    val textDim: String? = null,
    val ink: String? = null,
)

enum class PersonaMode(val raw: String) {
    GIF("gif"),
    TEXT("text");

    companion object {
        fun fromRaw(raw: String?): PersonaMode = when (raw) {
            "text" -> TEXT
            else -> GIF
        }
    }
}

@Serializable(with = StateFrames.Companion::class)
sealed class StateFrames {
    abstract val filenames: List<String>

    data class Single(val name: String) : StateFrames() {
        override val filenames: List<String> get() = listOf(name)
    }

    data class Variants(val names: List<String>) : StateFrames() {
        override val filenames: List<String> get() = names
    }

    data class Text(val frames: List<String>, val delayMs: Int) : StateFrames() {
        override val filenames: List<String> get() = emptyList()
    }

    companion object : KSerializer<StateFrames> {
        override val descriptor: SerialDescriptor =
            buildClassSerialDescriptor("StateFrames")

        override fun deserialize(decoder: Decoder): StateFrames {
            val element = (decoder as JsonDecoder).decodeJsonElement()
            return when (element) {
                is JsonPrimitive -> {
                    if (!element.isString) error("StateFrames primitive must be a string, got ${element.content}")
                    Single(element.content)
                }
                is JsonArray -> Variants(
                    element.map {
                        val p = it.jsonPrimitive
                        if (!p.isString) error("StateFrames array element must be a string")
                        p.content
                    }
                )
                is JsonObject -> {
                    val frames = element["frames"]?.jsonArray?.map {
                        val p = it.jsonPrimitive
                        if (!p.isString) error("frames element must be a string")
                        p.content
                    } ?: error("StateFrames object missing 'frames'")
                    val delay = element["delay_ms"]?.jsonPrimitive?.intOrNull
                        ?: element["delay"]?.jsonPrimitive?.intOrNull
                        ?: 200
                    Text(frames, delay)
                }
            }
        }

        override fun serialize(encoder: Encoder, value: StateFrames) {
            val jsonEncoder = encoder as JsonEncoder
            val element: JsonElement = when (value) {
                is Single -> JsonPrimitive(value.name)
                is Variants -> JsonArray(value.names.map { JsonPrimitive(it) })
                is Text -> buildJsonObject {
                    put("frames", JsonArray(value.frames.map { JsonPrimitive(it) }))
                    put("delay_ms", JsonPrimitive(value.delayMs))
                }
            }
            jsonEncoder.encodeJsonElement(element)
        }
    }
}

data class PersonaManifest(
    val name: String,
    val mode: PersonaMode,
    val colors: PersonaPalette,
    val states: Map<String, StateFrames>,
) {
    fun framesFor(slug: String): StateFrames? = states[slug]

    companion object {
        private val json = Json {
            ignoreUnknownKeys = true
            isLenient = true
        }

        fun fromJson(text: String): PersonaManifest? {
            val element = try {
                json.parseToJsonElement(text) as? JsonObject ?: return null
            } catch (_: Throwable) {
                return null
            }
            val name = element["name"]?.jsonPrimitive?.content ?: return null
            val mode = PersonaMode.fromRaw(element["mode"]?.jsonPrimitive?.content)

            val colors = element["colors"]?.let {
                runCatching { json.decodeFromJsonElement(PersonaPalette.serializer(), it) }.getOrNull()
            } ?: PersonaPalette()

            val states = element["states"]?.let { raw ->
                runCatching {
                    json.decodeFromJsonElement(
                        MapSerializer(String.serializer(), StateFrames),
                        raw,
                    )
                }.getOrNull()
            } ?: emptyMap()

            return PersonaManifest(name, mode, colors, states)
        }
    }
}
